### PR TITLE
CB-16918 - introduce retry for attached experience cluster fetch

### PR DESCRIPTION
**Pull request by gregito**

### CODE REVIEW

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private static final Logger LOCAL_LOGGER = LoggerFactory.getLogger(ExperienceCallUtil.class);
+
+    private ExperienceCallUtil() {
+    }
+
+    public static <T> Optional<T> fetchResultWithRetry(ResponseReader responseReader, URI path, Callable<Response> toCall, Class<T> resolveResponse,
+            String messageUponSingleFailure, Logger logger) {
+        throwIfNull(responseReader, () -> new IllegalArgumentException(ResponseReader.class.getName() + " instance should not be null!"));
+        throwIfNull(toCall, () -> new IllegalArgumentException("No " + Callable.class.getSimpleName() + " was given, therefore no action can be done!"));
+        Logger log = getIfNotNullOtherwise(logger, LOCAL_LOGGER);
```
[the remark this answers is not in the document]
Comment (biharitomi) [reply environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java:28]: I agree, I can not understand why the logger is a method parameter, could you please remove that too?

File: environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        byte retryCount = 0;
+        do {
+            try (Response response = executeCall(path, toCall, logger)) {
+                Optional<T> resolvedResponse = responseReader.read(path.toString(), response, resolveResponse);
```
Comment (biharitomi) [inline environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java:32]: logs are totally missing around this logics, debug and info logs preferably

File: common/src/main/java/com/sequenceiq/cloudbreak/util/NullUtil.java [hunk 3]
```diff
@@ -37,6 +50,13 @@ public static <T, R> R getIfNotNullOtherwise(T value, Function<T, R> consumer, R
         return toReturnIfValueIsNull;
     }
 
+    public static <R> R getIfNotNullOtherwise(R value, R toReturnIfValueIsNull) {
```
[the remark this answers is not in the document]
Comment (gregito) [reply common/src/main/java/com/sequenceiq/cloudbreak/util/NullUtil.java:53]: requireNonNull throws NPE when the second argument is null therefore these two methods are not the same

File: environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    private ExperienceCallUtil() {
+    }
+
+    public static <T> Optional<T> fetchResultWithRetry(ResponseReader responseReader, URI path, Callable<Response> toCall, Class<T> resolveResponse,
+            String messageUponSingleFailure, Logger logger) {
+        throwIfNull(responseReader, () -> new IllegalArgumentException(ResponseReader.class.getName() + " instance should not be null!"));
+        throwIfNull(toCall, () -> new IllegalArgumentException("No " + Callable.class.getSimpleName() + " was given, therefore no action can be done!"));
+        Logger log = getIfNotNullOtherwise(logger, LOCAL_LOGGER);
+        byte retryCount = 0;
```
Comment (biharitomi) [inline environment/src/main/java/com/sequenceiq/environment/experience/call/retry/ExperienceCallUtil.java:29]: Why custom retry is necessary? 
I assume the same could be accomplished via the project-wide `@Retryable` approaches.

**Review by biharitomi — CHANGES_REQUESTED** [review]

**Pull request closed without merging**